### DISCUSSION TITLE
How to access the environment variables when using createMiddleware()/createHandler()

### DISCUSSION THREAD

**Discussion by FraBle** [magne4000/universal-middleware#30] [General]
Example use case: configure an API client from environment variables.
For instance, Hono has an [`env()` adapter](https://hono.dev/docs/helpers/adapter#env) that makes the environment variables in Cloudflare accessible to a request handler/middleware via the context object.

createMiddleware()/createHandler() do accept parameters that can be passed to the middleware/handler, but expect them to be known at build time ([example from the docs](https://universal-middleware.dev/recipes/context-middleware)). In my use case, I'd like to pass in API client details from the environment, which lives in the context object in Hono, which is only known at runtime.

**Comment by magne4000** [comment]
An equivalent of Hono `env` could be a good idea, I need to think about it.

In the mean time, you can perhaps try to give an object with a function as your middleware parameters:
```ts
interface Arg {
  getEnv: () => Record<string, string>
}

const contextMiddleware = ((arg) => (request, ctx, runtime) => {
  const env = arg.getEnv();
  // ...
}) satisfies Get<[Arg], UniversalMiddleware>;
```

**Reply by magne4000** [reply]
You can also create an `env` helper like this:
```ts
export function env<T extends Record<string, unknown>>(runtime: RuntimeAdapter): T {
  switch (runtime.runtime) {
    case "bun":
    case "node":
    case "edge-light":
      return globalThis?.process?.env as T;
    case "deno":
      // @ts-ignore
      return Deno.env.toObject() as T;
    case "workerd":
      return (runtime.env ?? {}) as T;
    default:
      return {} as T;
  }
}
```

**Comment by magne4000** [comment]
[env helper](https://universal-middleware.dev/recipes/env-helper) available in `@universal-middleware/core@0.2.10` (same code as above)
